Attribute for configuring DeleteBehavior

Vekz opened the pull request:
Fixes #9621 

I've added:
DeleteBehaviorAttribute - Stores DeleteBehavior set on the property.
DeleteBehaviorAttributeConvention - Triggers on ForeignKeyAdded, checks if it's properties contain DeleteBehavior. If so then sets DeleteBehavior of this foreign key.
DeleteBehaviorAttributeConventionTest - Tests cases for setting all possible DeleteBehavior for explicit Foreign Key defined by attribute. Also checks for composite key, for multiple different keys with differernt Behaviors, for Foreign Key defined by FluentAPI and for implicit Foreign Key.

Moved:
DeleteBehavior => EFCore.Abstractions and added TypeForward in EFCore
(Also I had to add EFCore.Abstractions reference in base classes for Migrations and Snapshot tests)

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/efcore?pullRequest=27630) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/efcore?pullRequest=27630) before we can accept your contribution.<br/><br/>:x: Vekz [sign now](https://cla.dotnetfoundation.org/dotnet/efcore?pullRequest=27630)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/efcore?pullRequest=27630) it.</sub>

roji commented on src/EFCore/Metadata/Conventions/DeleteBehaviorAttributeConvention.cs:
```diff
@@ -0,0 +1,46 @@
+        var properties = foreignKey.Properties;
+        foreach (var property in properties)
+        {
+            var attribute = property?.PropertyInfo?.GetCustomAttribute<DeleteBehaviorAttribute>();
```
```suggestion
            var attribute = property.PropertyInfo?.GetCustomAttribute<DeleteBehaviorAttribute>();
```

roji reviewed (changes requested):
Here's a first round of comments, the others will likely have more.

Generally speaking, I think the delete behavior logically applies to the foreign key - or to the navigation - as a whole, and not on each property making up the foreign key. Specifically, the way this PR is currently implemented, it's possible to have multiple conflicting DeleteBehavior values on different properties making up the foreign key.

We should instead consider having the attribute on the navigation property from the dependent side. This would also support the case where [the foreign key is a shadow property](https://docs.microsoft.com/en-us/ef/core/modeling/relationships?tabs=fluent-api%2Cfluent-api-simple-key%2Csimple-key#no-foreign-key-property) (but let's see what @AndriySvyryd says before making this change).

Vekz commented on src/EFCore.Abstractions/DeleteBehavior.cs:
```diff
@@ -10,12 +10,15 @@ namespace Microsoft.EntityFrameworkCore;
 ///         Behaviors in the database are dependent on the database schema being created
-///         appropriately. Using Entity Framework Migrations or <see cref="DatabaseFacade.EnsureCreated" />
+///         appropriately. Using Entity Framework Migrations or
+///         <see href="https://docs.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.infrastructure.databasefacade.ensurecreated">
```
Of course it's better but I had to do it like that because of moving this Enum to EFCore.Abstractions as per @ajcvickers suggestion on the issue. - The classes couldn't be referenced from Abstractions.

roji replied:
You're right, I missed that.

However, it's probably better to remove those references altogether - the other types in EFCore.Abstractions simply don't reference types from the other packages. Rather than linking to DbContext, we can just use the word "context", and for APIs such as SaveChanges we can use `<c>SaveChanges</c>`.

AndriySvyryd replied:
For references that are important enough to be preserved as links remove `/en-us` from the URL

Vekz commented:
Okay, I'll fix those issues. 
Thank you @roji for being so thorough. Also that's great point on configuring DeleteBehavior on navigation properties.
I should've thrown an Exception in current solution.

AndriySvyryd commented:
> We should instead consider having the attribute on the navigation property from the dependent side. This would also support the case where [the foreign key is a shadow property](https://docs.microsoft.com/en-us/ef/core/modeling/relationships?tabs=fluent-api%2Cfluent-api-simple-key%2Csimple-key#no-foreign-key-property) (but let's see what @AndriySvyryd says before making this change).

I agree, FK properties could also be used be several FKs.

Vekz commented on test/EFCore.Tests/Metadata/Conventions/DeleteBehaviorAttributeConventionTest.cs:
```diff
@@ -0,0 +1,460 @@
+    }
+    #endregion
+    #region DeleteBehaviourAttribute set to Restrict and foreign key defined by FluentApi
+    private class Blog_Restrict_Fluent
```
I think I've done that, because each Blog type is pointing to the corresponding Post type through collection of them. I'm actually not sure should I change that. Although number of these classes will be smaller once I'm finished as per implementing your other suggestions.

Vekz commented:
I've done smaller formatting fixes and simplified convention tests.

Next time I'm gonna have some time I'll do the major change in convention to set attribute on navigation properties instead of parts of the key and the property attribute tests.

@AndriySvyryd How would you select which entries in xml docs are important enough to keep them as links? Is there some rule of thumb that I should keep in mind?

roji commented:
> How should I go about forcing user to set the attribute on the navigation property. Should I throw an exception if the attribute was set on the individual property of the foreign key instead of the navigation property itself (for now i just ignore that fact)? If so what kind of exception?

Yes, I'd vote for throwing an InvalidOperationException with the message `The [DeleteBehavior] attribute may only be specified on navigation properties, and is not supported not on properties making up the foreign key.`

> How would you select which entries in xml docs are important enough to keep them as links? Is there some rule of thumb that I should keep in mind?

The links here point to very general EF Core concepts (DbContext, SaveChanges) so I think they are quite low-value in general. Just mentioning "context" and `<c>SaveChanges</c>` should be fine.

ajcvickers commented:
@Vekz Have you completed all the requested changes such that this is ready for another review?

ajcvickers commented:
@Vekz No problem; just wanted to check you were not waiting on us.

Vekz commented:
I've also added the check for adding attribute on wrong side of the relationship. As @roji suggested it should work only from dependent side otherwise it will throw InvalidOperationException with the message `The [DeleteBehavior] attribute may only be specified on dependent side of the relationship.` Comments were also corrected.

Regarding (_overrides_configuration_from_convention_source, _does_not_override_configuration_from_explicit_source) tests, I guess as now the attribute is set on navigation they belong in `NavigationAttributeConventionTest.cs`, don't they?

AndriySvyryd commented on src/EFCore/Metadata/Conventions/DeleteBehaviorAttributeConvention.cs:
```diff
@@ -0,0 +1,63 @@
+
+        if (!navigationBuilder.Metadata.IsOnDependent)
+        {
+            throw new InvalidOperationException(CoreStrings.DeleteBehaviorAttributeOnPrincipalProperty);
```
For 1:1 relationships (when `foreignKey.IsUnique` is `true`) the dependent end could change after this is run. So just do nothing in that case and implement `IForeignKeyPrincipalEndChangedConvention` to set the `DeleteBehavior`. Also implement `IModelFinalizingConvention` to throw for invalid attributes on 1:1 relationships

Vekz replied:
What do you mean as invalid attribute on 1:1 relationship?

AndriySvyryd replied:
When the attribute is specified on the principal side

Vekz replied:
Ok. I think I've done that. Again sorry that takes me so long.

AndriySvyryd commented on src/EFCore/Metadata/Conventions/DeleteBehaviorAttributeConvention.cs:
```diff
@@ -55,9 +55,74 @@ public void ProcessNavigationAdded(IConventionNavigationBuilder navigationBuilde
         var foreignKey = navigationBuilder.Metadata.ForeignKey;
+        if (foreignKey.IsUnique)
+        {
+            return;
```
I may not have express myself clearly. I meant that this should stop processing when `!navigationBuilder.Metadata.IsOnDependent && foreignKey.IsUnique`

AndriySvyryd commented on src/EFCore/Metadata/Conventions/DeleteBehaviorAttributeConvention.cs:
```diff
@@ -0,0 +1,128 @@
+            return;
+        }
+        
+        if (foreignKey.GetDeleteBehaviorConfigurationSource() != ConfigurationSource.Explicit)
+        {
+            foreignKey.SetDeleteBehavior(navAttribute.Behavior);
+        }
```
```suggestion
        foreignKey.Builder.OnDelete(navAttribute.Behavior, fromDataAnnotation: true)
```

AndriySvyryd commented on src/EFCore/Metadata/Conventions/DeleteBehaviorAttributeConvention.cs:
```diff
@@ -0,0 +1,128 @@
+            return;
+        }
+        
+        if (relationshipBuilder.Metadata.GetDeleteBehaviorConfigurationSource() != ConfigurationSource.Explicit)
+        {
+            relationshipBuilder.Metadata.SetDeleteBehavior(navAttribute.Behavior);
+        }
```
```suggestion
        relationshipBuilder.OnDelete(navAttribute.Behavior, fromDataAnnotation: true)
```

AndriySvyryd commented on src/EFCore/Metadata/Conventions/DeleteBehaviorAttributeConvention.cs:
```diff
@@ -0,0 +1,128 @@
+            return;
+        }
+
+        var navigation = relationshipBuilder.Metadata.DependentToPrincipal;
+        if (navigation == null)
+        {
+            return;
+        }
+        
+        var navAttribute = navigation.PropertyInfo?.GetCustomAttribute<DeleteBehaviorAttribute>();
```
```suggestion
        var navigation = relationshipBuilder.Metadata.DependentToPrincipal;        
        var navAttribute = navigation?.PropertyInfo?.GetCustomAttribute<DeleteBehaviorAttribute>();
```

AndriySvyryd commented on test/EFCore.Tests/Metadata/Conventions/DeleteBehaviorAttributeConventionTest.cs:
```diff
@@ -0,0 +1,277 @@
+                    .Property(e => e.Blog_On_PrincipalId)).Message
+        );
+    }
+
```
Add a test where the attribute is on the principal side of a 1:1 relationship

Vekz commented:
@AndriySvyryd I think I've managed to make all your requested changes. Although at the end I merged main branch finally to keep up to date with repo and I think I've broke something as it broke heck a lot of imports. Gonna look into it closer in few days.

AndriySvyryd commented on test/EFCore.Tests/Metadata/Conventions/DeleteBehaviorAttributeConventionTest.cs:
```diff
@@ -0,0 +1,314 @@
+                modelBuilder.FinalizeModel()).Message
+        );
+    }
+
```
Remove empty line

AndriySvyryd commented on src/EFCore.Abstractions/DeleteBehaviorAttribute.cs:
```diff
@@ -0,0 +1,29 @@
+namespace Microsoft.EntityFrameworkCore;
+
+/// <summary>
+///     Configures the property or field to indicate how a delete operation is applied to dependent entities
```
```suggestion
///     Configures a navigation property to indicate how a delete operation is applied to dependent entities
```

AndriySvyryd reviewed (approved):
You still need to resolve the conflict in `CoreStrings.resx`, you can rebase on latest `main` and squash your changes. Also make sure to "Run Custom Tool" on `CoreStrings.Designer.tt` in VS

Vekz commented:
Shoot, I use Rider daily at my private PC. So you'll have to wait for me to install VS ;)

Vekz commented:
Okay, I've run the Custom Tool and I hope fixed tests that weren't passing on CIs.

AndriySvyryd commented:
Thanks for your contribution!

Vekz commented:
Thank you for your help with this issue. Especially that it was my first ever open source issue. It was really educative experience.
